### ISSUE TITLE
ISO 8601 support

### ISSUE THREAD

**Issue by ichernev** [moment/moment#2450]
You would think that moment supports ISO 8601 ... well only the pieces that we want to :)

- [x] support basic formats for date and time and date-time (no dashes)
- [x] support 2 digit timezone
- [ ] datetime separator can be `T`, or ` ` (space -- but not in strict mode) or nothing
- [x] fractional seconds can be split by `.` or `,`
- [ ] both `T` and `Z` are case insensitive. `t` and `z` are valid.

**Comment by mj1856** [comment]
Do you intend to also support the compact form?  I think currently we support only the extended form.

**Comment by ichernev** [comment]
@mj1865 that is called basic format, not compact (first item).

**Comment by mj1856** [comment]
got it.  thanks.

**Comment by AoDev** [comment]
Hi, I didn't know if opening a new issue or write here. But I found this behaviour for ISO a bit strange:

Trying in the console on momentjs.com:

```
moment().format()
> "2015-08-14T21:44:22+02:00"
```

```
moment().toISOString()
"2015-08-14T19:44:22.319Z"
```

The docs say

Default format

> As of version 1.5.0, calling moment#format without a format will default to moment.defaultFormat. Out of the box, moment.defaultFormat is the ISO8601 format YYYY-MM-DDTHH:mm:ssZ.

Shouldn't `toISOString()` and `format()` without format String be the same? (tested in Chrome 44)

**Comment by mj1856** [comment]
The characters in the `defaultFormat` string are all [display format specifiers](http://momentjs.com/docs/#/displaying/format/), including the `Z`.   The `Z` formatter returns the current UTC offset in +/-HH:mm format.

It's not the same as emitting a `'Z'` character in the output.

**Comment by AoDev** [comment]
Ah thanks, I get it.

**Comment by pietercolpaert** [comment]
Change regex in from-string.js:9 to:
```javascript
var isoRegex = /^\s*(?:[+-]\d{6}|\d{4})-?(?:(\d\d-\d\d)|(W\d\d$)|(W\d\d-?\d)|(\d\d\d))((T| )(\d\d(:?\d\d(:?\d\d([\.,]\d+)?)?)?)?([\+\-]\d\d(?::?\d\d)?|\s*Z)?)?$/;
```

**Comment by mj1856** [comment]
Both ISO8601 and RFC3339 allow `t` and `z` as lower case.  Added to the checklist.

**Comment by ichernev** [comment]
Fixed in #2641.
